src/group_theory/quaternion.lean

```diff
@@ -0,0 +1,281 @@
+/-
+Copyright (c) 2021 Julian Kuelshammer. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Julian Kuelshammer (heavily based on `dihedral` by Shing Tak Lam)
+-/
+import data.zmod.basic
+import group_theory.order_of_element
+import data.nat.basic
+import tactic.interval_cases
+
+/-!
+# Quaternion Groups
+
+We define the (generalised) quaternion groups `quaternion n` of order `4n`, also known as dicyclic
+groups, with elements `a i` and `xa i` for `i : zmod n`.
+
+## Main definition
+
+`quaternion n`: The (generalised) quaternion group of order `4n`.
+
+## Implementation notes
+
+`quaternion 0` is isomorphic to the infinite dihedral group, but it would be inconvenient to carry
+around the condition that `n` is greater than zero.
+
+## References
+
+https://en.wikipedia.org/wiki/Dicyclic_group
+https://en.wikipedia.org/wiki/Quaternion_group
+
+## TODO
+
+Show that `quaternion 2` is isomorphic to the subgroup of the quaternions generated by `i`, `j` and
+`k`.
+
+-/
+
+/--
+The (generalised) quaternion group `quaternion n` of order `4n`. It can be defined by the
+presentation `⟨a, x | a²ⁿ = 1, x² = aⁿ, x⁻¹ax=a⁻¹⟩`. We write `a i` for `aⁱ` and `xa i` for `xaⁱ`.
+-/
+@[derive decidable_eq]
+inductive quaternion (n : ℕ) : Type
```

Comment:
You may have to put these in a namespace or rename them, as the type of quaternions over a commutative ring is already called [`quaternion`](https://leanprover-community.github.io/mathlib_docs_demo/data/quaternion.html#quaternion).

Reply:
I don't mind. Any preference on whether to use `group.quaternion` or `quaternion_group`?

Reply:
I went for quaternion_group for now, but everyone feel free to suggest something different.

Reply:
That seems reasonable to me! The file should probably be renamed to `group_theory/quaternion_group` as well for consistency.

Reply:
Does it matter that that's inconsistent with `dihedral.lean`?

Reply:
Oh, I didn't realize that. My thought was that files should be named after the main declaration (if there is one), but I guess it could go either way then. Feel free to switch it back if you prefer!

Reply:
Looking through mathlib as a whole I feel like dihedral is the one sticking out.

Reply:
Feel free to rename dihedral in this PR if you like.

Reply:
Because of #6587 I am hesitant to make the rename here as it feels like git merge might mess such a big change up. I would prefer to do the name change either in the second of the two PRs or as a separate PR.

Reply:
Sounds good.